ipa-core/src/net/test.rs

```diff
@@ -10,33 +10,127 @@
 #![allow(clippy::missing_panics_doc)]
 use std::{
     array,
+    iter::zip,
     net::{SocketAddr, TcpListener},
 };
 
 use once_cell::sync::Lazy;
 use rustls_pki_types::CertificateDer;
 
-use super::transport::MpcHttpTransport;
+use super::{transport::MpcHttpTransport, ConnectionFlavor, Shard};
 use crate::{
     config::{
         ClientConfig, HpkeClientConfig, HpkeServerConfig, NetworkConfig, PeerConfig, ServerConfig,
         TlsConfig,
     },
     executor::{IpaJoinHandle, IpaRuntime},
-    helpers::{HandlerBox, HelperIdentity, RequestHandler},
+    helpers::{repeat_n, HandlerBox, HelperIdentity, RequestHandler, TransportIdentity},
     hpke::{Deserializable as _, IpaPublicKey},
     net::{ClientIdentity, Helper, MpcHelperClient, MpcHelperServer},
+    sharding::ShardIndex,
     sync::Arc,
     test_fixture::metrics::MetricsHandle,
 };
 
-pub const DEFAULT_TEST_PORTS: [u16; 3] = [3000, 3001, 3002];
+/// Simple struct to keep default port configuration organized.
+pub struct Ports<C> {
+    ring: C,
+    shards: C,
+}
+
+/// A **single** ring with 3 hosts, each with a ring and sharding port.
+pub const DEFAULT_TEST_PORTS: Ports<[u16; 3]> = Ports {
+    ring: [3000, 3001, 3002],
+    shards: [6000, 6001, 6002],
```

Comment:
how many shard does this setup support? it is not clear to me why we need sharding ports open on a single MPC ring and any other configuration with more than one shard per helper cannot be supported by this configuration it seems

Reply:
1) how many shard does this setup support? 1 shard. It's an "unsharded" scenario. At this point I'm only adding facilities to the code. Exercising more shards will come later. This to make PRs simpler.

2) "it is not clear to me why we need sharding ports open on a single MPC ring"
We spoke about this sometime back. The reason is simplicity. Having shards open ports all the time allows for the code to not have to bother for the special case of one shard.

3) "any other configuration with more than one shard per helper cannot be supported by this configuration it seems"
`DEFAULT_TEST_PORTS` is just a default for a single ring MPC. I was planning on adding one for sharded at some point later.

Reply:
I am not sure whether simplicity savings here are justified, especially for in-memory tests, as it invites us to setup listeners and other heavy stuff for these ports that otherwise won't be used and complexity is contained to this module.

If you think it will take too long to fix, I am ok proceeding with this but at least we need a comment explaining what is going on here